File: workspace/lib/xod/core/flip-flop/any.cpp
struct State {
};

{{ GENERATED_CODE }}

void evaluate(Context ctx) {
    bool oldState = getValue<output_MEM>(ctx);
    bool newState = oldState;

    if (isInputDirty<input_TGL>(ctx)) {
        newState = !oldState;
    } else if (isInputDirty<input_SET>(ctx)) {
        newState = true;
    } else {
        newState = false;
    }

    if (newState == oldState)
        return;

    emitValue<output_MEM>(ctx, newState);
}

File: workspace/lib/xod/core/flip-flop/any.js
module.exports.setup = function(e) {
  e.context.state = false;
};
module.exports.evaluate = function(e) {
  var state = e.context.state;
  var newState;

  if (e.inputs.TGL) {
    newState = !e.context.state;
  } else if (e.inputs.SET) {
    newState = true;
  } else {
    newState = false;
  }

  if (newState === state) return;

  e.context.state = newState;
  return { MEM: newState };
};
